server/@tsplang-plugins/scripts/generate-enums.js

```diff
@@ -0,0 +1,806 @@
+                function(enum)
+                    if enum == nil then return ""; end;
+
+                    if enum["__tostring"] ~= nil then
+                        local result = tostring(enum);
+
+                        if enum["__tonumber"] ~= nil then
+                            result = result.."\\t"..tostring(tonumber(enum));
+                        end;
+
+                        return result.."\\n";
+                    end;
+
+                    return "";
+                end;
+            ),
+            __mtable = getmetatable(smu.ON),
+            __consecutive_nils = 0,
+            __i = 0,
+            next = (
+                function(self)
+                    while (self.__i < 16777215) do
+                        local enumtable = self.__mtable[self.__i];
+                        self.__i = self.__i + 1;
+
+                        if enumtable ~= nil then
+                            self.__consecutive_nils = 0;
+                            local packet = "";
+
+                            local zeroth = enumtable[0];
+                            if zeroth ~= nil then
+                                packet = self.__make_entry(zeroth);
+                            end;
+
+                            for j,enum in ipairs(enumtable) do
+                                packet = packet..self.__make_entry(enum);
+                            end;
+
+                            if packet ~= "" then
+                                return tostring(self.__i - 1).."\\n"..packet;
+                            end;
+                        else
+                            self.__consecutive_nils = self.__consecutive_nils + 1;
+                        end;
+
+                        if self.__consecutive_nils > 1000000 then
+                            break;
+                        end;
+                    end;
+
+                    return nil;
+                end;
+            ),
+            done = (
+                function(self)
+                    localnode.prompts = self.prevprompts;
+                    _G["SCRAPER"] = nil;
+                    return nil;
+                end;
+            )
+        };
+    `.replace(/\s{2,}|\n/gm, " "),
+
+    /**
+     * @memberof InstrumentComms
+     * @param {string} ip
+     * @param {number} port
+     * @returns {Promise<Array<string>>} Returns the data received from the
+     * instrument on Promise resolution.
+     */
+    connect: function(ip, port) {
+        return new Promise((resolve, reject) => {
+            let cleaned = false
+            /** @type {Array<string>} */
+            const received = []
+
+            const socket = new net.Socket({
+                readable: true,
+                writable: true,
+            })
+            socket.once("connect", function() {
+                console.log(`-> Connected to ${ip}:${port}`)
+                socket.write(InstrumentComms.SCRAPER + "\n")
+            })
+            socket.on("data", function(data) {
```

Comment:
When you write `InstrumentComms.SCRAPER` to the instrument, will the instrument always send back something that will trigger the `socket.on("data",..)` event listener? I didn't see another `socket.write(...)` call outside of connect.

Reply:
You're right. I had set up my test server incorrectly.
I'll add a `print(SCRAPER:next())` command after the Lua table declaration.